chain-vote/src/private_voting/unit_vector_zkp.rs

```diff
@@ -0,0 +1,459 @@
+        challenge_x: &Scalar,
+        challenge_y: &Scalar,
+    ) -> bool {
+        let bits = ciphertexts.bits();
+        let length = ciphertexts.len();
+        let cx_pow = challenge_x.power(bits);
+
+        let powers_cx = challenge_x.exp_iter();
+        let powers_cy = challenge_y.exp_iter();
+
+        let powers_z_iterator = powers_z_encs_iter(&self.zwvs, challenge_x, &(bits as u32));
+
+        let zero = public_key.encrypt_with_r(&Scalar::zero(), &self.r);
+
+        let mega_check = GroupElement::multiscalar_multiplication(
+            self.zwvs.iter().map(|zwv| zwv.z)
+                .chain(self.zwvs.iter().map(|zwv| zwv.w))
+                .chain(iter::repeat(challenge_x.negate()).take(bits))
+                .chain(iter::repeat(Scalar::one().negate()).take(bits))
+                .chain(self.zwvs.iter().map(|zwv| zwv.v))
+                .chain(self.zwvs.iter().map(|zwv| zwv.z - challenge_x))
+                .chain(iter::repeat(Scalar::one().negate()).take(bits))
+                .chain(powers_cy.take(length).map(|s| s * cx_pow))
+                .chain(powers_cy.take(length).map(|s| s * cx_pow))
+                .chain(powers_cy.take(length).map(|s| s))
+                .chain(powers_cx.take(bits).map(|s| s))
+                .chain(powers_cx.take(bits).map(|s| s))
+                .chain(iter::once(Scalar::one().negate()))
+                .chain(iter::once(Scalar::one().negate())),
+            iter::repeat(GroupElement::generator()).take(bits)
+                .chain(iter::repeat(commitment_key.h).take(bits))
+                .chain(self.ibas.iter().map(|iba| iba.i))
+                .chain(self.ibas.iter().map(|iba| iba.b))
+                .chain(iter::repeat(commitment_key.h).take(bits))
+                .chain(self.ibas.iter().map(|iba| iba.i))
+                .chain(self.ibas.iter().map(|iba| iba.a))
+                .chain(ciphertexts.iter().map(|ctxt| ctxt.e2))
+                .chain(ciphertexts.iter().map(|ctxt| ctxt.e1))
+                .chain(powers_z_iterator.take(length).map(|p| p))
+                .chain(self.ds.iter().map(|ctxt| ctxt.e1))
+                .chain(self.ds.iter().map(|ctxt| ctxt.e2))
+                .chain(iter::once(zero.e1))
+                .chain(iter::once(zero.e2)),
+        );
+
+        mega_check == GroupElement::zero()
+    }
+
+    /// Final verification of the proof. We do not use the multiscalar optimisation when using sec2 curves.
+    #[cfg(not(feature = "ristretto255"))]
+    fn verify_statements(
+        &self,
+        public_key: &PublicKey,
+        commitment_key: &CommitmentKey,
+        ciphertexts: &Ptp<Ciphertext>,
+        challenge_x: &Scalar,
+        challenge_y: &Scalar,
+    ) -> bool {
+
+        // check commitments are 0 / 1
+        for (iba, zwv) in self.ibas.iter().zip(self.zwvs.iter()) {
+            let com1 = commitment_key.commit_with_random(&zwv.z, &zwv.w);
+            let lhs = &iba.i * challenge_x + &iba.b;
+            if lhs != com1 {
```

Comment:
Could this (and the following check) be rewritten as `commitment_key.verify(iba.i * challenge_x + &iba.b, Open {m: zwv.z, r: zwv.w })`?
If so, `commit_with_random` could be made even private

Reply:
Very good catch 👌